crates/amaru-ledger/src/rules/block/body_size.rs

```diff
@@ -0,0 +1,93 @@
+use crate::protocol_parameters::ProtocolParameters;
+use crate::rules::RuleViolation;
+use amaru_kernel::{cbor, MintedBlock};
+pub struct BlockBodySizeTooBig {
+    pub supplied: usize,
+    pub max: usize,
+}
+
+impl From<BlockBodySizeTooBig> for RuleViolation {
+    fn from(value: BlockBodySizeTooBig) -> Self {
+        RuleViolation::BlockBodySizeTooBig(value)
+    }
+}
+
+pub fn block_body_size_valid(
+    block: &MintedBlock<'_>,
+    protocol_params: &ProtocolParameters,
+) -> Result<(), BlockBodySizeTooBig> {
+    let max_block_size = protocol_params
+        .max_block_body_size
+        .try_into()
+        .unwrap_or_else(|_| panic!("Failed to convert u32 to usize"));
+
+    let mut tx_bodies_raw = Vec::new();
+    if cbor::encode(block.transaction_bodies.clone(), &mut tx_bodies_raw).is_err() {
+        // TODO: handle appropriately
+        panic!("failed to encode block transaction_bodies in CBOR");
+    }
+
+    let mut tx_witness_sets_raw = Vec::new();
+    if cbor::encode(
+        block.transaction_witness_sets.clone(),
+        &mut tx_witness_sets_raw,
+    )
+    .is_err()
+    {
+        // TODO: handle appropriately
+        panic!("failed to encode block transaction_witness_sets in CBOR");
+    }
+
+    let mut auxiliary_data_set_raw = Vec::new();
+    if cbor::encode(
+        block.auxiliary_data_set.clone(),
+        &mut auxiliary_data_set_raw,
+    )
+    .is_err()
+    {
+        // TODO: handle appropriately
+        panic!("failed to encode block auxiliary_data_set in CBOR");
+    }
+
+    let mut invalid_tranasctions_raw = Vec::new();
+    if cbor::encode(
+        block.invalid_transactions.clone(),
+        &mut invalid_tranasctions_raw,
+    )
+    .is_err()
+    {
+        // TODO: handle appropriately
+        panic!("failed to encode block invalid_transactions in CBOR");
+    }
+
+    let block_body_size = tx_bodies_raw.len()
+        + tx_witness_sets_raw.len()
+        + auxiliary_data_set_raw.len()
+        + invalid_tranasctions_raw.len();
```

Comment:
This really ought to be `block.raw_cbor().len()` or something like that. We shouldn't be re-serializing and cloning structures here. 

Although, the Haskell implementation seems a little bit contrived about how the block size is computed  (using some "abstract sizing" based on types) . We need to ensure the calculations are the same and not only that an arbitrarily chosen block is below the limit. 

If anything, leave a `// FIXME` at the top of this function with details about what's up and need to be done.

Reply:
Totally agreed. It seems the block sizing is everything but the header, so I can't use the whole raw CBOR, but yes. This is sloppy and risky.

Reply:
The check in the Haskell code is only validating that the purported size in the header is within the bounds, not that the actual block size is valid. I would think the actual check to make sure the purported size is accurate would happen during serialization in Pallas (though it doesn't seem to happen currently).

I'll update this to check the header instead of the actual bytes.